src/stream/src/executor/backfill/snapshot_backfill.rs

```diff
@@ -476,59 +443,47 @@ impl<'a, S: UpstreamBufferState> UpstreamBuffer<'a, S> {
                 .ok_or_else(|| anyhow!("end of upstream"))?;
             match msg {
                 DispatcherMessage::Chunk(chunk) => {
+                    self.is_polling_epoch_data = true;
                     self.consume_upstream_row_count
                         .inc_by(chunk.cardinality() as _);
                 }
                 DispatcherMessage::Barrier(barrier) => {
-                    self.state.on_upstream_barrier(barrier);
-                    break Ok(());
+                    self.is_polling_epoch_data = false;
+                    break Ok(barrier);
+                }
+                DispatcherMessage::Watermark(_) => {
+                    self.is_polling_epoch_data = true;
                 }
-                DispatcherMessage::Watermark(_) => {}
             }
         }
     }
 }
 
-impl<'a, 's> UpstreamBuffer<'a, StateOfConsumingLogStore<'s>> {
-    async fn take_buffered_barrier(&mut self) -> StreamExecutorResult<Option<Barrier>> {
-        Ok(if let Some(barrier) = self.state.barriers.pop_back() {
-            Some(barrier)
-        } else if !self.state.upstream_pending_barriers.is_empty() {
-            let barrier = self.state.handle_one_pending_barrier().await?;
-            Some(barrier)
-        } else if self.state.is_finished {
-            None
-        } else {
-            self.consume_until_next_barrier().await?;
-            let barrier = self.state.handle_one_pending_barrier().await?;
-            Some(barrier)
-        })
-    }
-}
-
-fn is_finish_barrier(
-    barrier: &Barrier,
-    current_subscriber_id: u32,
-    upstream_table_id: TableId,
-) -> bool {
-    if let Some(Mutation::DropSubscriptions {
-        subscriptions_to_drop,
-    }) = barrier.mutation.as_deref()
-    {
-        let is_finished = subscriptions_to_drop
-            .iter()
-            .any(|(subscriber_id, _)| *subscriber_id == current_subscriber_id);
-        if is_finished {
-            assert!(subscriptions_to_drop.iter().any(
-                |(subscriber_id, subscribed_upstream_table_id)| {
-                    *subscriber_id == current_subscriber_id
-                        && upstream_table_id == *subscribed_upstream_table_id
+impl<'a> UpstreamBuffer<'a, StateOfConsumingLogStore> {
+    async fn next_barrier(&mut self) -> StreamExecutorResult<Option<DispatcherBarrier>> {
+        Ok(
+            if let Some(barrier) = self.state.upstream_pending_barriers.pop_back() {
+                // only update the `max_pending_barrier_num` on checkpoint barrier to avoid deadlock caused by
+                // downstream waiting on a checkpoint barrier to bump up.
```

Comment:
What do you mean by "downstream waiting on a checkpoint barrier to bump up"? This comment seems to suggest that updating `max_pending_barrier_num` will somehow result in downstream waiting for checkpoint barrier, but I'm not so certain why.

Some comments would be helpful here.

Reply:
Same scenario as https://github.com/risingwavelabs/risingwave/pull/18798/files#r1795161729 above.

Reply:
Updated the comment. The reason is the same as https://github.com/risingwavelabs/risingwave/pull/18798#discussion_r1795161729.